### PR TITLE
ILM: Make all the shrink action steps retryable

### PR DESCRIPTION
**Pull request by andreidan**
This aims at making the shrink action retryable. Every step is
retryable, but in order to provide an experience where ILM tries
to achieve a successful shrink even when the target node goes
missing permanently or the shrunk index cannot recover, this also
introduces a retryable shrink cycle within the shrink action.

The shrink action will generate a unique index name that'll be the
shrunk index name. The generated index name is stored in the lifecycle
state.

If the shrink action ends up waiting for the source shards to
collocate or for the shrunk index to recover for more than the configured
`LIFECYCLE_STEP_WAIT_TIME_THRESHOLD` setting, it will move back
to clean up the attempted (and failed) shrunk index and will retry
generating a new index name and attempting to shrink the source
to the newly generated index name.

Relates to #48183

### CODE REVIEW

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java [hunk 1]
```diff
@@ -47,6 +48,13 @@
     public static final Setting<TimeValue> LIFECYCLE_STEP_MASTER_TIMEOUT_SETTING =
         Setting.positiveTimeSetting(LIFECYCLE_STEP_MASTER_TIMEOUT, TimeValue.timeValueSeconds(30), Setting.Property.Dynamic,
             Setting.Property.NodeScope);
+    // This setting configures how much time since step_time should ILM wait for a condition to be met. After the threshold wait time has
+    // elapsed ILM will likely stop waiting and go to the next step.
+    // Also see {@link org.elasticsearch.xpack.core.ilm.ClusterStateWaitUntilThresholdStep}
+    public static final Setting<TimeValue> LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING =
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:54]: Would we want to document this?

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java [hunk 2]
```diff
@@ -650,20 +653,6 @@ private void putShrinkOnlyPolicy(RestClient client, String policyName) throws IO
                         builder.endObject();
                     }
                     builder.endObject();
-
-                    // Sometimes throw in an extraneous unfollow just to check it doesn't break anything
-                    if (randomBoolean()) {
```
Comment (dakrone) [inline x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java:655]: Why was this removed?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java [hunk 3]
```diff
@@ -47,6 +48,13 @@
...
+        Setting.positiveTimeSetting(LIFECYCLE_STEP_WAIT_TIME_THRESHOLD, TimeValue.timeValueHours(12), Setting.Property.Dynamic,
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:55]: Should we enforce a minimum setting? It would be bad to be spinning forever because someone put 1 second as their wait time.

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for working on this Andrei! I left some comments for an initial review.

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        ClusterState.Builder newClusterStateBuilder = ClusterState.builder(clusterState);
+
+        LifecycleExecutionState lifecycleState = fromIndexMetadata(indexMetadata);
+
+        Builder newCustomData = LifecycleExecutionState.builder(lifecycleState);
+        String policy = indexMetadata.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+        String generatedIndexName = generateValidIndexName(prefix, index.getName());
+        ActionRequestValidationException validationException = validateGeneratedIndexName(generatedIndexName, clusterState);
+        if (validationException != null) {
+            logger.warn("unable to generate a valid shrink index name as part of policy [{}] for index [{}] due to [{}]",
```
[the remark this answers is not in the document]
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java:84]: Ah yes, good catch

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java (hunk 2 above)
Comment (andreidan) [reply x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java:655]: Ah, I think I was suspecting this could cause some flakiness, but I believe it might've been due to dev issues as I was working through this feature (saving the execution state - namely the shrink index name being present in the execution state when we reach the cold pohase). I'll add this back. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java (hunk 3 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:55]: Good point. Shall we set it to one hour minimum? (this will make it impossible to integration test - https://github.com/elastic/elasticsearch/pull/70107/files/db8a13789e40a0b7f64263a0c88a257b21eb206c#diff-00728663bc72dd25e9832acf55fd6c4dd690a34d3c9ef6e5408464d81ff1f785R309 - however, maybe the unit tests we have around the `ClusterStateWaitUntilThresholdStep` are enough)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ClusterStateWaitUntilThresholdStep.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        Result stepResult = stepToExecute.isConditionMet(index, clusterState);
+
+        if (stepResult.isComplete() == false) {
+            // checking the threshold after we execute the step to make sure we execute the wrapped step at least once (because time is a
+            // wonderful thing)
+            TimeValue retryThreshold = LifecycleSettings.LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING.get(idxMeta.getSettings());
+            LifecycleExecutionState lifecycleState = fromIndexMetadata(idxMeta);
+            if (waitedMoreThanThresholdLevel(retryThreshold, lifecycleState, Clock.systemUTC())) {
```
[the remark this answers is not in the document]
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ClusterStateWaitUntilThresholdStep.java:72]: Oh, you're absolutely right - wall clocks are not trustworthy. I don't think `nanoTime` would be the answer and the reading invariant on that api is to be monotonically increasing, more so than accuracy.

I don't think this is too bad because if we get a reading where the time drift is hours away we'll retry the shrink cycle (so nothing truly bad happens, except a wasteful - and presumbaly eventually successful - shrink operation, with the note that we'll at least delete the abandoned shrunk index)

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java (hunk 3 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java]: I managed to integration test the rewind even with a `1h` minimum value in https://github.com/elastic/elasticsearch/pull/70107/commits/73e363dec50e3773d90390f1e50582a55fa05b3b

**Comment by andreidan** [comment]
@dakrone thanks for the review. I think I addressed all your comments

**Comment by andreidan** [comment]
@jrodewig can you please help us with the docs section (I had an initial go at documenting the new behaviour of the shrink action)

**Review by jrodewig — COMMENTED** [review]
Thank @andreidan. I pushed 0fa3ca2d08adc25054984511762fdf876af6e679 to refactor the docs a bit.

That commit moves the info about allocation and retrying steps to the bottom of the ILM shrink page. I feel that retrying steps should be relatively rare. Most users probably don't need to worry about it.

I also added some reference docs for `index.lifecycle.step.wait_time_threshold`. Please feel free to make further changes or reach out if you have questions. Thanks again!

File: docs/reference/ilm/actions/ilm-shrink.asciidoc [hunk 6]
```diff
@@ -103,3 +91,25 @@ PUT _ilm/policy/my_policy
...
+==== Retrying failed allocation steps
+
+During a `shrink` action, {ilm-init} allocates the source index's primary shards
+to one node. After shrinking the index, {ilm-init} reallocates the shrunken
+index's shards to the appropriate nodes based on your allocation rules. These
+allocation steps can fail if:
+
+* A node is removed during the `shrink` action.
+* No node has enough disk space to host the source index's shards.
+* {es} cannot reallocate the shrunken index due to conflicting allocation rules.
```
Comment (andreidan) [inline docs/reference/ilm/actions/ilm-shrink.asciidoc:105]: @jrodewig these are some examples of the possible conditions that might not be fulfilled. Shall we rephrase this a bit to illustrate that? (as opposed to being the only possible issues)

File: docs/reference/ilm/actions/ilm-shrink.asciidoc [hunk 7]
```diff
@@ -103,3 +91,25 @@ PUT _ilm/policy/my_policy
   }
 }
 --------------------------------------------------
+
+[[ilm-shrink-retry-failed-allocation]]
+==== Retrying failed allocation steps
```
Comment (andreidan) [inline docs/reference/ilm/actions/ilm-shrink.asciidoc:96]: @jrodewig the threshold and retry-ability/rewind of the shrink action is based on waiting for some conditions that are not fulfilled until the threshold is breached/surpassed. Should we be focused more on the "wait until threshold and then rewind" as opposed to retrying a step?

File: docs/reference/settings/ilm-settings.asciidoc [hunk 8]
```diff
@@ -59,6 +59,15 @@ An index that was rolled over would normally match the full format,
 for example `logs-2016.10.31-000002`). 
 If the index name doesn't match the pattern, index creation fails.
 
+[[index-lifecycle-step-wait-time-threshold]]
+`index.lifecycle.step.wait_time_threshold`::
+(<<indices-update-settings,Dynamic>>, <<time-units,time value>>) 
+Time to wait while retrying a failed allocation step for the {ilm-init}
```
Comment (andreidan) [inline docs/reference/settings/ilm-settings.asciidoc:65]: @jrodewig I'm not sure if we should use "while retrying" as we are not necessarily retrying steps, but rather waiting for conditions to be fulfilled (ie. source index shards be collocated, shrunk index shards be recovered/allocated)

File: docs/reference/settings/ilm-settings.asciidoc [hunk 9]
```diff
@@ -59,6 +59,15 @@ An index that was rolled over would normally match the full format,
...
+<<ilm-shrink,`shrink`>> action. After this period, {ilm-init} re-attempts the
```
Comment (andreidan) [inline docs/reference/settings/ilm-settings.asciidoc:66]: @jrodewig "re-attempts the previous step" is not entirely accurate as the shrink action rewinds back a few steps, depending on the situation. Maybe we shouldn't be so prescriptive here and say the `shrink` action (which is the only place where this is used now) "does the right thing"

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java [hunk 10]
```diff
@@ -177,14 +180,33 @@ public boolean isSafeAction() {
         CheckNotDataStreamWriteIndexStep checkNotWriteIndexStep = new CheckNotDataStreamWriteIndexStep(checkNotWriteIndex,
             waitForNoFollowerStepKey);
         WaitForNoFollowersStep waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, readOnlyKey, client);
-        UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, setSingleNodeKey, client, readOnlySettings);
+        UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, cleanupShrinkIndexKey, client, readOnlySettings);
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:183]: While we're changing this step, can you make this use `ReadOnlyStep` instead of the generic settings update one?

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks Andrei, I left some more comments. I also played around with this manually and found an interesting thing. If we don't use a number of shards (using `max_primary_shard_size` instead), then if manually moving an index that is *past* a shrink step back to any point *before* the shrink, the cleanup step actually ends up deleting the index itself!

I think it'd be great if we could figure out a way to avoid this, maybe in the `CleanupShrinkIndexStep` we can check to see if the index name matches the pre-existing name in `lifecycleExecutionState.getShrinkIndexName()` then we skip doing the shrink? What do you think? (I know this is an edge case, I was just curious to see how it behaved)

File: docs/reference/ilm/actions/ilm-shrink.asciidoc (hunk 6 above)
Comment (jrodewig) [reply docs/reference/ilm/actions/ilm-shrink.asciidoc:106]: Thanks @andreidan. I rephrased as follows:
> These allocation steps can fail for a number of reasons, including:

File: docs/reference/ilm/actions/ilm-shrink.asciidoc (hunk 7 above)
Comment (jrodewig) [reply docs/reference/ilm/actions/ilm-shrink.asciidoc]: Good point. I renamed this "Shard allocation for shrink" as it could also be useful for people curious about how shrink allocates normally.

File: docs/reference/settings/ilm-settings.asciidoc (hunk 8 above)
Comment (jrodewig) [reply docs/reference/settings/ilm-settings.asciidoc]: I reworded to:
> Time to wait for the cluster to resolve allocation issues during an ILM `shrink` action.

File: docs/reference/settings/ilm-settings.asciidoc (hunk 9 above)
Comment (jrodewig) [reply docs/reference/settings/ilm-settings.asciidoc]: This is a good point. We don't really need to document the remedial action here anyway. That's already covered in the xref.

**Comment by jrodewig** [comment]
Thanks for the review and iteration @andreidan. I pushed a few more edits to the docs with 2ac08e741a0a7d8831d0d0fea0ebe9a0306a3f8c. Happy to address any other feedback you have!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * This generates a valid unique index name by using the provided prefix, appended with a generated UUID, and the index name.
+     */
+    static String generateValidIndexName(String prefix, String indexName) {
+        String randomUUID = generateValidIndexSuffix(() -> UUIDs.randomBase64UUID().toLowerCase(Locale.ROOT));
+        randomUUID = randomUUID.substring(0, Math.min(randomUUID.length(), MAX_GENERATED_UUID_LENGTH));
+        return prefix + randomUUID + "-" + indexName;
+    }
+
+    static String generateValidIndexSuffix(Supplier<String> randomGenerator) {
```
[the remark this answers is not in the document]
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java:145]: I pondered that too, but I still think the suffix is still a better name as it might not be valid if it'd be used as a prefix (ie. we don't make sure the generated UUID doesn't *start with* `_` or `-`)

**Comment by andreidan** [comment]
@jrodewig thanks so much for restructuring and rewording the docs

**Comment by andreidan** [comment]
> if manually moving an index that is past a shrink step back to any point before the shrink, the cleanup step actually ends up deleting the index itself

@dakrone ah interesting. So this happened because the alias/identity of the source index was swapped to the shrink index (and the source index was deleted in the process). I pushed https://github.com/elastic/elasticsearch/pull/70107/commits/4b091e683b71ce57c8de63b90a8d3da014aa4868 to skip the cleanup index step if the managed index is a shrunk index and the source index does not exist anymore

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ExecuteStepsUpdateTask.java [hunk 12]
```diff
@@ -127,6 +124,10 @@ public ClusterState execute(final ClusterState currentState) throws IOException
                     } catch (Exception exception) {
                         return moveToErrorStep(state, currentStep.getKey(), exception);
                     }
+                    // some steps can decide to change the next step to execute after waiting for some time for the condition
+                    // to be met (eg. {@link LifecycleSettings#LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING}, so it's important we
+                    // re-evaluate what the next step is after we evaluate the condition
+                    nextStepKey = currentStep.getNextStepKey();
```
[the remark this answers is not in the document]
Comment (andreidan) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ExecuteStepsUpdateTask.java:130]: Great catch! Removed `nextStepKey` from the logging statement as it's logged below lines 132+

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java:145]: Okay, that's fine with me :)

**Comment by andreidan** [comment]
@elasticmachine update branch

**Review by dakrone — APPROVED** [review]
LGTM, thanks for iterating on this Andrei!

**Review by jrodewig — APPROVED** [review]
